core/distributed/index_map.cpp

```diff
@@ -0,0 +1,187 @@
+
+#include <ginkgo/core/distributed/index_map.hpp>
+
+
+#include "core/distributed/index_map_kernels.hpp"
+
+
+namespace gko {
+namespace index_map_kernels {
+
+
+GKO_REGISTER_OPERATION(build_mapping, index_map::build_mapping);
+GKO_REGISTER_OPERATION(map_to_local, index_map::map_to_local);
+
+
+}  // namespace index_map_kernels
+
+
+namespace experimental {
+namespace distributed {
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+size_type index_map<LocalIndexType, GlobalIndexType>::get_local_size() const
+{
+    return partition_ ? partition_->get_part_size(rank_) : 0;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+size_type index_map<LocalIndexType, GlobalIndexType>::get_non_local_size() const
+{
+    return remote_global_idxs_.get_size();
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+const segmented_array<GlobalIndexType>&
+index_map<LocalIndexType, GlobalIndexType>::get_remote_global_idxs() const
+{
+    return remote_global_idxs_;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+const segmented_array<LocalIndexType>&
+index_map<LocalIndexType, GlobalIndexType>::get_remote_local_idxs() const
+{
+    return remote_local_idxs_;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+const array<comm_index_type>&
+index_map<LocalIndexType, GlobalIndexType>::get_remote_target_ids() const
+{
+    return remote_target_ids_;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+std::shared_ptr<const Executor>
+index_map<LocalIndexType, GlobalIndexType>::get_executor() const
+{
+    return exec_;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+size_type index_map<LocalIndexType, GlobalIndexType>::get_global_size() const
+{
+    return partition_ ? partition_->get_size() : 0;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+array<LocalIndexType> index_map<LocalIndexType, GlobalIndexType>::map_to_local(
+    const array<GlobalIndexType>& global_ids, index_space index_space_v) const
+{
+    array<LocalIndexType> local_ids(exec_);
+
+    exec_->run(index_map_kernels::make_map_to_local(
+        partition_.get(), remote_target_ids_, to_device(remote_global_idxs_),
+        rank_, global_ids, index_space_v, local_ids));
+
+    return local_ids;
+}
+
+
+template <typename LocalIndexType, typename GlobalIndexType>
+index_map<LocalIndexType, GlobalIndexType>::index_map(
+    std::shared_ptr<const Executor> exec,
+    std::shared_ptr<const partition_type> partition, comm_index_type rank,
+    const array<GlobalIndexType>& recv_connections)
+    : exec_(std::move(exec)),
+      partition_(std::move(partition)),
+      rank_(rank),
+      remote_target_ids_(exec_),
+      remote_local_idxs_(exec_),
+      remote_global_idxs_(exec_)
+{
+    array<LocalIndexType> flat_remote_local_idxs(exec_);
+    array<GlobalIndexType> flat_remote_global_idxs(exec_);
+    array<int64> remote_sizes(exec_);
```

Comment:
can be 
```suggestion
    array<GlobalIndexType> remote_sizes(exec_);
```
?

Reply:
The segmented array always uses int64 offsets, and the sizes have to be the same type.